### PR TITLE
Move openfermion functions to correct modules

### PR DESCRIPTION
**Pull request by soranjh**
**Context:**
The qchem functions that depend on openfermion and pyscf are all collected in the `openfermion_pyscf` module. 

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by soranjh** [comment]
[sc-59835]

File: pennylane/qchem/dipole.py [hunk 1]
```diff
@@ -320,3 +320,189 @@ def _dipole(*args):
...
+            qml.qchem.dipole_moment(mol, cutoff=cutoff, core=core, active=active, mapping=mapping)(
+                *(args or [])
+            )
+            if requires_grad
+            else qml.qchem.dipole_moment(
+                mol, cutoff=cutoff, core=core, active=active, mapping=mapping
+            )()
+        )
+        if wires:
+            wires_new = qml.qchem.convert._process_wires(wires)
```
Comment (obliviateandsurrender) [inline pennylane/qchem/dipole.py:486]: Need to disable pylint's `protected-access` warning here.

File: pennylane/qchem/dipole.py [hunk 2]
```diff
@@ -320,3 +320,189 @@ def _dipole(*args):
...
+        return dip
+
+    dip = qml.qchem.dipole_of(
+        symbols,
+        geometry_hf,
+        molecule.name,
+        molecule.charge,
+        molecule.mult,
+        molecule.basis_name,
+        package="pyscf",
```
Comment (obliviateandsurrender) [inline pennylane/qchem/dipole.py:499]: I thought we supported computing dipole moments with `openfermion` backend as well. 🤔

Comment (soranjh) [reply pennylane/qchem/dipole.py:499]: Yes, molecular_dipole has two backends: dhf and openfermion. The openfermion backend calls the dipole_of function which uses openfermion, and openfermion internally calls pyscf for computing integrals. That is why dipole_of has  `package="pyscf"`. We should depricate dipole_of eventually and make it a private function.

**Review by obliviateandsurrender — APPROVED** [review]
Looks good to go! Don't forget to add it to the changelog [here](https://github.com/PennyLaneAI/pennylane/blob/fb06ce880a6e6942eed268cdcb03fb8c06403643/doc/releases/changelog-dev.md?plain=1#L290).

**Review by ddhawan11 — APPROVED** [review]
Looks good.

**Pull request merged**